DIPs/DIP1xxx-ajie.md

```diff
@@ -0,0 +1,222 @@
+{   auto result = DynamicArray!(ElementType!R).init;
+    foreach(el; range) result.put(el);
+    return result;
+}
+
+void writeDeepLengthA(Roi)(Roi rangeOfIterables)
+{   typeof(rangeOfIterables.front.front) sum;
+    foreach(iterable; rangeOfIterables) sum += iterable.length;
+    sum.writeln;
+}
+
+void writeDeepLengthB(Roi)(Roi rangeOfIterables)
+{   typeof(rangeOfIterables.front.front) sum;
+    foreach(ref iterable; rangeOfIterables) sum += iterable.length;
+    sum.writeln;
+}
+
+//Enabled by this DIP
+void writeDeepLengthC(Roi)(Roi rangeOfIterables)
+{   typeof(rangeOfIterables.front.front) sum;
+    foreach(auto ref iterable; rangeOfIterables) sum += iterable.length;
+    sum.writeln;
+}
+
+void main(){
+    // Elements of this range can be copied but have no address
+    auto unaddressedElements = iota(0, 10).map!(i => iota(0, i));
+    // Vice-versa
+    auto uniqueElements =
+    [   only(7, 2, 29, 30).dynamicArray,
+        only(11, 9, 0).dynamicArray,
+        takeNone!(int[]).dynamicArray,
+        only(3, 30, 14, 19, 4, 0, 9).dynamicArray
+    ];
+
+    // Ok, prints 45
+    unaddressedElements.writeDeepLengthA;
+    // Error: struct `containers.dynamicarray.DynamicArray!(int, Mallocator, false).DynamicArray`
+    // is not copyable because it is annotated with @disable
+    uniqueElements.writeDeepLengthA;
+
+    // Will be error after DMD pull request 8437 [1] is merged
+    unaddressedElements.writeDeepLengthA;
+    // Ok, prints 14
+    uniqueElements.writeDeepLengthB;
+
+    // Ok, prints 45
+    unaddressedElements.writeDeepLengthC;
+    // Ok, prints 14
+    uniqueElements.writeDeepLengthC;
+}
+```
+
+## Alternatives
+
+- The plan to disable by-reference iteration of rvalue ranges [1] could be cancelled.
+    This has the disadvantage that the purpose of the loop will become harder to see,
+    because one cannot assume that `ref` means that the loop excepts reference semantics.
+    On the other hand, if DIP1016 [4] gets accepted, this might be the most consistent
+    alternative with rest of the language.
+
+- Programmers could be intstructed to use introspection to select whether to iterate by
+    reference or by value. This will make coding general-purpose non-mutating loops a
+    lot more difficult compared to this proposal, and greatly decreases the likelihood
+    of third-party code accepting ranges with non-copyable members.
+
+- `for` loops could be instructed to be used instead of `foreach`. The disadvantages
+    are same as above.
+
+- A library solution could be made that takes an `alias` compile-time parameter and
+    chooses the iteration method on behalf of the programmer.
+    `std.algorithm.iteration.each` [5] would be a good canditate to become one. This
+    concept has the following disadvantages:
+    - Error messages become harder to read than with normal loops,
+    - Using `goto`, labeled `break` and `continue`, and `return` inside the loop to exit
+        it becomes impossible.
```

Comment:
*impossible*? No. It can be worked around, but in general, it's quite hard. We have the compiler to handle that.